Mathlib/Algebra/Order/GroupWithZero/WithZero.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2024 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+import Mathlib.Algebra.Order.GroupWithZero.Canonical
+import Mathlib.Algebra.Order.GroupWithZero.Unbundled
+/-!
+
+# Covariant instances on `WithZero`
+
+Adding a zero to a type with a preorder and multiplication which satisfies some
+axiom, gives us a new type which satisfies some variant of the axiom.
+
+## Example
+
+If `α` satisfies `b₁ < b₂ → a * b₁ < a * b₂` for all `a`,
+then `WithZero α` satisfies `b₁ < b₂ → a * b₁ < a * b₂` for all `a > 0`,
+which is `PosMulStrictMono (WithZero α)`.
+
+## Application
+
+The type `ℤₘ₀ := WithZero (Multiplicative ℤ)` is used a lot in mathlib's valuation
+theory. These instances enable lemmas such as `mul_pos` to fire on `ℤₘ₀`.
+
+-/
+
+assert_not_exists Ring
```

Comment:
Is this really supposed to survive in mathlib once the file compiles?

Reply:
Yes, the point is to check that this file does not import `Ring`.

Reply:
Well, but I understand this while preparing the file. If it imports it, CI would give an error, no?

Reply:
My understanding is that if someone adds an import to some other file at some later time, and this happens to end up with this file importing Ring, then we discover this before the import addition is merged.